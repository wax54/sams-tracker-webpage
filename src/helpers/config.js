export const timeFrames = {
    day: {
        title: "Every Day",
        val: 1,
        roundPlaces: 2
    },
    week: {
        title: "Every Week",
        val: 7,
        roundPlaces: 1
    },
    month: {
        title: "Every Month",
        val: 30,
        roundPlaces: 0
    }
}