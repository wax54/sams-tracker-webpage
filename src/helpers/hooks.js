import { useState, useEffect } from 'react';

function getWindowDimensions() {
    const { innerWidth: width, innerHeight: height } = window;
    return {
        width,
        height
    };
}

export function useWindowDimensions() {
    const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

    useEffect(() => {
        function handleResize() {
            setWindowDimensions(getWindowDimensions());
        }

        window.addEventListener('resize', handleResize);
        return () => window.removeEventListener('resize', handleResize);
    }, [setWindowDimensions]);

    return windowDimensions;
}


export function useFormFields(init={}) {
    const [formData, setFormData] = useState(init);
    
    const handleChange = evt => {
        const {name, value} = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
    };
    const resetForm = () => setFormData(init);
    return [formData, handleChange, resetForm];
}