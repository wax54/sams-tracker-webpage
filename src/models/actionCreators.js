import UserApi from "./Api";
import {v4 as uuid} from "uuid";
import UploadQueue from "./UploadQueue";
import { store } from "./store";
import { SET_TIMEFRAME } from "./actionTypes";

/** Timeframes */
export function setTimeFrame(timeframeKey) {
    return { type: SET_TIMEFRAME, payload: timeframeKey };
}

/** SHIFTS */
export function startShift(type, category) {
    const tempId = uuid();
    const newShift = {id: tempId, start: new Date(), type, category};
    
    // add a command and shift/data to the upload queue
    UploadQueue.addToQueue({type:"addShift", data: newShift});
    //add the shift to the store
    return { type: "START_SHIFT", payload: newShift };
}

export function resetShifts() {
    return { type: "RESET_SHIFTS" };
}

export function loadShifts() {
    return async function (dispatch) {
        const resp = await UserApi.getShifts();
        if (resp.status === true) {
            dispatch({ type: "ADD_SHIFTS", payload: resp.shifts });
            return true;
        }
        if (resp.status === false) {
            console.error(resp.errors);
            return false;
            //TODO make a queue for failed shift uploads to be merged later
            //dispatch({ type: "ADD_SHIFT_TO_UPLOAD_QUEUE", payload: newShift })
        }
    }
}

export function refreshShifts() {
    return async function (dispatch) {
        dispatch(resetShifts());
        return await dispatch(loadShifts());
    }
}

export function updateShift(shift) {
    return { type: "UPDATE_SHIFT", payload: shift };
}

export function updateAShift(shift) {
    UploadQueue.addToQueue({ type: "updateShift", data: shift });
    return updateShift(shift);
}

export function endShift(id) {
    const stop = new Date();
    return clockOutAt(id, stop);
}

export function clockOutAt(shiftId, stop) {
    const shift = store.getState().shifts[shiftId];
    shift.stop = stop;
    UploadQueue.addToQueue({type: 'updateShift', data: shift})
    return updateShift(shift);
}

export function deleteShift(shiftId) {
    UploadQueue.addToQueue({type: 'deleteShift', data: shiftId});
    return { type: "DELETE_SHIFT", shiftId: shiftId };
}

/** ERRORS (depreciated) */
export function resetErrors() {
    return { type: "RESET_ERRORS" };
}


/** Users */
export function setUser(userData) {
    return { type: "SET_USER", payload: userData };
}
export function resetUser() {
    return async function (dispatch) { 
        dispatch({ type: "RESET_USER" });
        dispatch(resetShifts());
        dispatch(resetGoals());
    }
}
export function authorizeUser({ username, password }) {
    return async function (dispatch) {
        const resp = await UserApi.login({username, password});
        if(resp.status === true)
            dispatch(setUser(resp.user));
            await dispatch(refreshShifts());
            await dispatch(refreshGoals());

        return resp;
    }
}

export function registerUser({ username, password }) {
    return async function (dispatch) {
        const resp = await UserApi.register({ username, password });
        if (resp.status === true)
            dispatch(setUser(resp.user));
            dispatch(resetShifts());
            dispatch(resetGoals());
        return resp;
    }
}

/** GOALS */
export function addGoal({type, category, seconds_per_day}) {
    const newGoal = { seconds_per_day, type, category };
    UploadQueue.addToQueue({type: "addGoal", data: newGoal});
    return {type: "ADD_GOAL", payload: newGoal};
}

export function resetGoals() {
    return { type: "RESET_GOALS" };
}

export function loadGoals() {
    return async function (dispatch) {
        const resp = await UserApi.getGoals();
        console.log(resp);
        if (resp.status === true) {
            dispatch({ type: "LOAD_GOALS", payload: resp.goals });
            return true;
        }
        if (resp.status === false) {
            console.error(resp.errors);
            return false;
        }
    }
}

export function refreshGoals() {
    return async function (dispatch) {
        dispatch(resetGoals());
        return await dispatch(loadGoals());
    }
}

export function updateGoal({type, category, seconds_per_day}) {
    return { type: "UPDATE_GOAL", payload: {type, category, seconds_per_day} };
}

export function updateAGoal(goal) {
    UploadQueue.addToQueue({type: "updateGoal", data: goal});
    return updateGoal(goal);
}

export function deleteGoal({ type, category }) {
    UploadQueue.addToQueue({type: "deleteGoal", data: { type, category }})
    return { type: "DELETE_GOAL", payload: { category, type } };
}