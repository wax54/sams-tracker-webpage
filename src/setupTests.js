// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';


export const MockApi = {
    CONNECTION_ERROR : "DB CONNECTION ERROR",
    VALID_CREDENTIALS : {username: "John Smith", password: "I AM GREAT"},
    VALID_USER : {
        id: 0, username: "John Smith", join_at: new Date(), last_login_at: new Date()
    },
    API_URL: "",
    token: false,
    
        //    { status: true, data: resp.data[type] };
        //    { status: false, errors };

    //USERS
    async login({ username, password }) {
        if(username === this.VALID_CREDENTIALS.username &&
            password === this.VALID_CREDENTIALS.password) {
            this.token = true;
            return {status: true, user:this.VALID_USER};
        } else {
            this.token = false;
            return { status: false, errors: [`Invalid Credentials!`] };
        }
    },

    async register({ username, password }) {
        if (username === this.VALID_CREDENTIALS.username &&
            password === this.VALID_CREDENTIALS.password) {
            this.token = true;
            return { status: true, user: this.VALID_USER };
        } else {
            this.token = false;
            return { status: false, errors: [`Invalid Credentials!`] };
        }
    },

    //SHIFTS
    getShifts: async function() {
        // if (!this.token) return { status: false, errors: ["USER NOT LOGGED IN"] };
        const shifts = [];
        return { status: true, shifts };
    },


    async addShift(shift) {
        const resp = await this.request({
            method: "post",
            url: this.API_URL + "/shifts/",
            data: { shift }
        });
        if (resp.status === false) return resp;
        else return { status: true, shift: resp.data.shift };
    },
    async updateShift(id, shift) {
        const resp = await this.request({
            method: "patch",
            url: this.API_URL + `/shifts/${id}`,
            data: { shift }
        });
        if (resp.status === false) return resp;
        return { status: true, shift: resp.data.shift };

    }
    ,async clockOutShift(id, time) {
        const resp = await this.request({
            method: "patch",
            url: this.API_URL + `/shifts/${id}`,
            data: { shift: { stop: time } }
        });
        if (resp.status === false) return resp;
        return { status: true, shift: resp.data.shift };

    },
    async deleteShift(id) {
        const resp = await this.request({
            method: "delete",
            url: this.API_URL + `/shifts/${id}`
        });
        if (resp.status === false) return resp;
        return { status: true };
    },

    //GOALS
    async getGoals() {
        const goals = [];
        return { status: true, goals: goals };

    },
    async addGoal(goal) {
        const resp = await this.request({
            method: "post",
            url: this.API_URL + "/goals",
            data: { goal }
        });
        if (resp.status === false) return resp;
        return { status: true, goal: resp.data.goal };
    },
    async updateGoal({ category, type }, seconds_per_day) {

        const resp = await this.request({
            method: "patch",
            url: this.API_URL + "/goals/",
            data: { goal: { seconds_per_day, category, type } }
        });
        if (resp.status === false) return resp;
        return { status: true, goal: resp.data.goal };

    },
    async removeGoal({ category, type }) {
        const resp = await this.request({
            method: "delete",
            url: this.API_URL + `/goals/`,
            data: { goal: { type, category } }
        });
        if (resp.status === false) return resp;
        return { status: resp.data.deleted };
    }

}

jest.mock('./models/Api', () => ({ ...MockApi }));
