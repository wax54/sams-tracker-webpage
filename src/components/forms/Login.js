import './Login.css';

import React, {useState} from "react";
import {useDispatch} from 'react-redux';
// import { useHistory } from "react-router";
// import UserContext from "../UserContext";
import { authorizeUser} from "../../models/actionCreators";
import SimpleForm from "./SimpleForm";
import { useHistory } from 'react-router';
import { useFormFields } from '../../helpers/hooks';
import { Link } from 'react-router-dom';

const LoginForm = () => {

    const loginInputs = {
        username: "",
        password: ""
    }

    const history = useHistory();
    const dispatch = useDispatch(); 
    const [errors, setErrors] = useState([]);
    const [passwordVisible, toggleVisible] = useState(false);

    const [ { username, password }, handleChange ] = useFormFields(loginInputs);


    const login = async evt => {
        evt.preventDefault();
        const userStatus = await dispatch(authorizeUser({ username, password }));
        if (userStatus.status) {
            history.push('/');
        } else {
            setErrors([userStatus.errors]);
        }
    }

    return (
        <form className="Login container-fluid" onSubmit={login} >
            {errors.map(message =>
                <div key={message} className="alert alert-danger">
                    {message}
                </div>
            )}
            <div className="input-group input-group-lg my-3">

                <label
                    htmlFor="login-password"
                    className="input-group-text">  Username: </label>
                <input
                    type="text"
                    className="form-control"
                    name="username"
                    id="login-username"
                    value={username}
                    onChange={handleChange}
                />
            </div>

            <div className="input-group input-group-lg my-3">
                <label 
                    htmlFor="login-password" 
                    className="input-group-text"> 
                    Password: 
                </label>


                <input
                    type={passwordVisible ? "text" : "password"}
                    className="form-control"

                    name="password"
                    id="login-password"
                    value={password}
                    onChange={handleChange}
                />
                <label className="input-group-text btn btn-secondary" htmlFor="signup-password" style={{}} onClick={() => toggleVisible(curr => !curr)} >
                    {passwordVisible ? "Hide" : "View"}
                </label>
            </div>
            <input className="btn btn-lg btn-block btn-primary" type="submit" value="Login!" />
            <Link className="btn btn-block btn-secondary" to="/signup">Need to Sign Up?</Link>

    </form>
    )};
export default LoginForm