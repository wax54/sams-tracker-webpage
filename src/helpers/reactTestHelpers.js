import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"



function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

export function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

export async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}
