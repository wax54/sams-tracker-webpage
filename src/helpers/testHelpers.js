import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"

function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}



export {
    renderWithRedux,
    renderAndWaitFor
}