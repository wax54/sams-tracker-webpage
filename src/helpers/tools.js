import { ShiftCollection, Shift, daysFrom } from "../models/ShiftCollection";
export function makeColor (shift) {
    const str = shift.type + shift.category;
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash = str.charCodeAt(i) + ((hash << 5) - hash);
    }
    let colour = '#';
    for (let i = 0; i < 3; i++) {
        const value = (hash >> (i * 8)) & 0xFF;
        colour += ('00' + value.toString(16)).substr(-2);
    }
    return colour;
}

export function round(num, decimals = 0) {
    if (decimals < 0) {
        decimals = decimals * -1;
        const divider = 10 ** decimals;
        num = Math.round(num / divider) * divider;
    }
    const multiplier = 10 ** decimals;
    return Math.round(num * multiplier) / multiplier;
}

// shifts 
        // {
        //     _hours: 17.3,
        //     _categories: ["school", "my well being"],
        //     _currShifts:[shift],
        //     school: { 
        //         _hours: 12, 
        //         _types:["coding"], 
        //         coding: { 
        //             _hours: 12, 
        //             _shifts: [shift, shift, shift] 
        //         }
        //     },
        //     "my well being": {
        //         _hours: 5.3,
        //         _types: ["resting", "painting"],
        //         resting: {
        //             _hours: 2.7,
        //             _shifts: [shift,  shift]
        //         },
        //         painting: {
        //             _hours: 2.6,
        //             _shifts: [shift]
        //         }
        //     },
        //  }

export function getShiftsByCategory(storeShifts, timeFrame) {
    const shifts = {
        _allShifts: function () {
            const shifts = new ShiftCollection();
            this._categories.entries().forEach(cat => {
                this[cat]._types.forEach(type => {
                    this[cat][type]._shifts.forEach(shift => {
                        shifts.add(shift);
                    });
                });
            });
            return shifts;
        },
        _currShifts: [],
        _hours: 0,
        _categories: new Set()
    };
    const earliestDateInTimeFrame = daysFrom(-timeFrame.val);

    for (let key in storeShifts) {
        const shift = new Shift(storeShifts[key]);
        //only shows shifts in timeFrame
        if (shift.stop < earliestDateInTimeFrame) continue;

        const hours = shift.getHours();
        if (!shift.stop) shifts._currShifts.push(shift);
        //if there is no entry for this category...
        if (!shifts._categories.has(shift.category)) {
            //make it
            shifts._categories.add(shift.category);
            shifts[shift.category] = { _hours: 0, _types: new Set() };
        }
        //shifts[shift.category] is garunteed at this point
        //if there is no entry for this type in the category...
        if (!shifts[shift.category]._types.has(shift.type)) {
            // make it
            shifts[shift.category]._types.add(shift.type);
            shifts[shift.category][shift.type] = { _hours: 0, _shifts: [] };
        }
        shifts[shift.category][shift.type]._shifts.push(shift);
        shifts[shift.category][shift.type]._hours += hours;
        shifts[shift.category]._hours += hours;
        shifts._hours += hours;
    }
    return shifts;
}