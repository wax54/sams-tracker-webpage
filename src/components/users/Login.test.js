import { renderWithRedux, renderAndWaitFor } from '../../helpers/reactTestHelpers';
import Login from './Login';
import { createStore } from 'redux';
import userEvent from '@testing-library/user-event';
import { MockApi } from '../../setupTests';
import { store } from '../../models/store';
import { waitFor } from '@testing-library/react';
import { resetAll } from '../../models/actionCreators';

//the history object we will be seeing if called
const mockHistory = { push: jest.fn() };
jest.mock('react-router-dom', () => (
    {
        ...jest.requireActual('react-router-dom'),
        useHistory: () => mockHistory,
    }));



beforeEach(() => store.dispatch(resetAll()));

test("doesn't blow up", () => {
    renderWithRedux(Login);
});

test("matchesSnapshot", () => {
    const { asFragment } = renderWithRedux(Login);
    expect(asFragment()).toMatchSnapshot();
});

test("inputs on page", () => {
    const { getByLabelText } = renderWithRedux(Login);

    const usernameInput = getByLabelText("Username:");
    expect(usernameInput).toBeInTheDocument();
    const passwordInput = getByLabelText("Password:");
    expect(passwordInput).toBeInTheDocument();
});

test("Can Type into Inputs", async () => {

    const { getByLabelText, getByText, debug } = renderWithRedux(Login);

    const usernameInput = getByLabelText("Username:");

    const passwordInput = getByLabelText("Password:");
    
    userEvent.type(usernameInput, "HELLO");
    userEvent.type(passwordInput, "GOODBYE" );

    expect(usernameInput.value).toEqual("HELLO");
    expect(passwordInput.value).toEqual("GOODBYE");

});

test("Submit works with valid creds", async () => {
    createStore
    const { getByLabelText, getByText } = renderWithRedux(Login,);

    const usernameInput = getByLabelText("Username:");
    const passwordInput = getByLabelText("Password:");
    userEvent.type(usernameInput, MockApi.VALID_CREDENTIALS.username);
    userEvent.type(passwordInput, MockApi.VALID_CREDENTIALS.password);

    const loginBtn = getByText("Login!");

    userEvent.click(loginBtn);

    await waitFor(() => expect(mockHistory.push).toBeCalledWith('/'));
    expect(store.getState().user).toEqual(MockApi.VALID_USER);

});

test("Submit fails with invalid creds", async () => {
    createStore
    const { getByLabelText, getByText, debug } = renderWithRedux(Login,);

    const usernameInput = getByLabelText("Username:");
    const passwordInput = getByLabelText("Password:");
    userEvent.type(usernameInput, MockApi.VALID_CREDENTIALS.username);
    userEvent.type(passwordInput, "BAD PASSWORD");

    const loginBtn = getByText("Login!");

    userEvent.click(loginBtn);

    await waitFor(() => expect(mockHistory.push).not.toBeCalledWith('/'));
    expect(store.getState().user).toEqual({});
    expect(getByText(`Invalid Credentials!`)).toBeInTheDocument();

});