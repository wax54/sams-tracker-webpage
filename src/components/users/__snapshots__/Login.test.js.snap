// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchesSnapshot 1`] = `
<DocumentFragment>
  <form
    class="Login container-fluid"
  >
    <div
      class="input-group input-group-lg my-3"
    >
      <label
        class="input-group-text"
        for="login-username"
      >
          Username: 
      </label>
      <input
        class="form-control"
        id="login-username"
        name="username"
        type="text"
        value=""
      />
    </div>
    <div
      class="input-group input-group-lg my-3"
    >
      <label
        class="input-group-text"
        for="login-password"
      >
        Password:
      </label>
      <input
        class="form-control"
        id="login-password"
        name="password"
        type="password"
        value=""
      />
      <label
        class="input-group-text btn btn-secondary"
        for="signup-password"
      >
        View
      </label>
    </div>
    <input
      class="btn btn-lg btn-block btn-primary"
      type="submit"
      value="Login!"
    />
    <a
      class="btn btn-block btn-secondary"
      href="/signup"
    >
      Need to Sign Up?
    </a>
  </form>
</DocumentFragment>
`;
